File: examples/governance/model_registry_domain/lib/model_registry_domain/data_usage_agreement/data_usage_agreement.rb
require 'hecks/model'

module ModelRegistryDomain
  class DataUsageAgreement
    autoload :Restriction, "model_registry_domain/data_usage_agreement/restriction"
    autoload :Lifecycle, "model_registry_domain/data_usage_agreement/lifecycle"

    include Hecks::Model

    attribute :model_id
    attribute :data_source
    attribute :purpose
    attribute :consent_type
    attribute :effective_date
    attribute :expiration_date
    attribute :restrictions, default: [], freeze: true
    attribute :status

    # State predicates — see lifecycle.rb for full state machine
    def draft?; status == "draft"; end
    def active?; status == "active"; end
    def revoked?; status == "revoked"; end

    VALID_CONSENT_TYPE = ["public_domain", "CC-BY-SA", "licensed", "consent", "opt-out"].freeze unless defined?(VALID_CONSENT_TYPE)

    private

    def validate!
      raise ValidationError, "data_source can't be blank" if data_source.nil? || (data_source.respond_to?(:empty?) && data_source.empty?)
      raise ValidationError, "purpose can't be blank" if purpose.nil? || (purpose.respond_to?(:empty?) && purpose.empty?)
      if consent_type && !VALID_CONSENT_TYPE.include?(consent_type)
        raise ValidationError, "consent_type must be one of: #{VALID_CONSENT_TYPE.join(', ')}, got: #{consent_type}"
      end
    end

    def check_invariants!
      raise InvariantError, "expiration must be after effective date" unless instance_eval(&proc { !expiration_date || !effective_date || expiration_date.to_s >= effective_date.to_s })
    end
  end
end

File: examples/governance_static_go/domain/data_usage_agreement.go
package domain

import (
	"time"
	"github.com/google/uuid"
	"fmt"
)

type DataUsageAgreement struct {
	ID string `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	ModelId string `json:"model_id"`
	DataSource string `json:"data_source"`
	Purpose string `json:"purpose"`
	ConsentType string `json:"consent_type"`
	EffectiveDate time.Time `json:"effective_date"`
	ExpirationDate time.Time `json:"expiration_date"`
	Restrictions []Restriction `json:"restrictions"`
	Status string `json:"status"`
}

func NewDataUsageAgreement(modelId string, dataSource string, purpose string, consentType string, effectiveDate time.Time, expirationDate time.Time, restrictions []Restriction, status string) *DataUsageAgreement {
	a := &DataUsageAgreement{
		ID:        uuid.New().String(),
		ModelId: modelId,
		DataSource: dataSource,
		Purpose: purpose,
		ConsentType: consentType,
		EffectiveDate: effectiveDate,
		ExpirationDate: expirationDate,
		Restrictions: restrictions,
		Status: status,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *DataUsageAgreement) Validate() error {
	if a.DataSource == "" {
		return &ValidationError{Field: "data_source", Message: "data_source can't be blank"}
	}
	if a.Purpose == "" {
		return &ValidationError{Field: "purpose", Message: "purpose can't be blank"}
	}
	if a.ConsentType != "" {
		validConsentType := map[string]bool{"public_domain": true, "CC-BY-SA": true, "licensed": true, "consent": true, "opt-out": true}
		if !validConsentType[a.ConsentType] {
			return &ValidationError{Field: "consent_type", Message: fmt.Sprintf("consent_type must be one of: public_domain, CC-BY-SA, licensed, consent, opt-out, got: %s", a.ConsentType)}
		}
	}
	// invariant: expiration must be after effective date
	return nil
}
